enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,307 @@ package hit_tracker_client
+	// local cache hits to the RPC service at the configured backend.
+	return &HitTrackerClient{ctx: ctx, enqueueFn: h.enqueue, client: h.client, requestMetadata: requestMetadata}
 }
 
-// TODO(iain): hook this up with a to-be-added RPC service in the app.
-type HitTracker struct {
+type NoOpHitTracker struct{}
+
+func (h *NoOpHitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+}
+
+func (h *NoOpHitTracker) TrackMiss(d *repb.Digest) error {
+	return nil
 }
 
-func (h HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (h *NoOpHitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (h *NoOpHitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h *HitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+type NoOpTransferTimer struct {
 }
 
-func (h *HitTracker) TrackMiss(d *repb.Digest) error {
+func (t *NoOpTransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
 	return nil
 }
 
+func (t *NoOpTransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
+	return nil
+}
+
+type HitTrackerClient struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	client          hitpb.HitTrackerServiceClient
+	requestMetadata *repb.RequestMetadata
+}
+
+// This should only be used for Action Cache hit tracking.
+func (h *HitTrackerClient) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+	alert.UnexpectedEvent("Unexpected call to SetExecutedActionMetadata")
+}
+
+// Local cache misses hit the backing cache, which will take care of
+// hit-tracking for this request.
+func (h *HitTrackerClient) TrackMiss(d *repb.Digest) error {
+	return nil
+}
+
+func (h *HitTrackerFactory) groupID(ctx context.Context) groupID {
+	claims, err := claims.ClaimsFromContext(ctx)
+	if err != nil {
+		return interfaces.AuthAnonymousUser
+	}
+	return groupID(claims.GetGroupID())
+}
+
+func (h *HitTrackerFactory) enqueue(ctx context.Context, requestMetadata *repb.RequestMetadata, hit *hitpb.CacheHit) {
+	requestMetadata = proto.Clone(requestMetadata).(*repb.RequestMetadata)
+
+	if h.stopped.Load() {
+		log.Warning("hit_tracker_client.enqueue after worker shutdown, sending RPC synchronously")
+		hit.RequestMetadata = requestMetadata
+		if _, err := h.client.Track(ctx, &hitpb.TrackRequest{Hits: []*hitpb.CacheHit{hit}}); err != nil {
+			log.Infof("Error sending HitTrackerService.Track RPC: %v", err)
+		}
+		return
+	}
+
```

Comment:
if `stopped` transitions from true -> false on this line, it looks like we'll buffer hits that will never be flushed.

In `redisutil.CommandBuffer` we use this pattern

```
	c.mu.Lock()
	if c.shouldFlushSynchronously() {
		c.mu.Unlock()
		return c.rdb.IncrBy(ctx, key, increment).Err()
	}
	defer c.mu.Unlock()

  // (... add to buffer ...)
```

Reply:
Oops, nice catch. Switched to match the code you pasted.